Collect basic hive table statistics on table write

arhimondr opened the pull request:
`rowCount`, `fileCount`, `rawSize`, `totalSize`

kokosing commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveBasicStatistics.java:
```diff
@@ -81,6 +83,38 @@ public String toString()
+        else if (!second.isPresent()) {
+            return first;
+        }
+        else {
```
redundant else

kokosing commented on presto-product-tests/src/main/java/com/facebook/presto/tests/hive/TestHiveBasicTableStatistics.java:
```diff
@@ -0,0 +1,400 @@
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+@Requires(ImmutableNationTable.class)
+public class TestHiveBasicTableStatistics
```
can you add tests with EXTERNAL non empty tables without stats, partitioned and non partitioned. What is happening when you do an INSERT to such tables?

arhimondr replied:
If writes to non managed tables is disabled (`hive.non-managed-table-writes-enabled`) exception is raised. If writes are enabled it should do the same way as for managed tables. 

This PR does not cover the case when statistics recomputing is needed. Like when you are inserting into an existing table with no statistics computed for the data it already has. For such cases the real rows count, and rows count in statistics will be different.

findepi replied:
> If we are inserting into existing table/partition, and statistics are not present - than throw an exception

you don't need to throw in that case. You can simply skip saving stats.

arhimondr replied:
Yeah, that's a good idea. I haven't thought about that. Will do.

arhimondr replied:
@findepi done

kokosing replied:
>> If we are inserting into existing table/partition, and statistics are not present - than throw an exception

>you don't need to throw in that case. You can simply skip saving stats.

Do you have tests for this case?

arhimondr replied:
I added unit tests. Can add product tests as well.

arhimondr replied:
UPD.

Turns out Hive always sets statistics, even on `CREATE EXTERNAL TABLE`. Those are just zero if the table is empty. 

Also Hive doesn't allow you to overwrite table statistics parameters by doing `ALTER TABLE nation SET TBLPROPERTIES ('totalSize' = 1234);`

The only way to reset those statistics is to directly connect to the metastore and override those parameters. Since connecting to metastore and changing something directly is not quite a "product test" i just decided to implement an integration test.

findepi commented on presto-hive/src/test/java/com/facebook/presto/hive/AbstractTestHiveClient.java:
```diff
@@ -2611,6 +2694,19 @@ private void doInsertIntoNewPartition(HiveStorageFormat storageFormat, SchemaTab
             existingFiles = listAllDataFiles(transaction, tableName.getSchemaName(), tableName.getTableName());
             assertFalse(existingFiles.isEmpty());
+
+            // test statistics
```
Is `TestHiveClient` run as a product test?

Can we have more product tests? Something like `INSERT ..` / `CTAS` followed by `SHOW STATS` ? These tests can be optionally both transactional and non-transactional.
(like we have for `SHOW PARTITIONS` --  `TestShowPartitions`)

arhimondr replied:
`AbstractTestHiveClient` is a set of `Metastore <-> Hive Connector` integration tests. We run them for all types of supported Metastores, including `FileMetastore`, `HiveMetastore` and FB internal Metastore. So adding a product test won't cover all the cases. From the other side i don't think duplicating test cases is something worth doing. 

Product tests for the statistics were supposed to verify that the real `Hive` can read them, and that the fields used to store them are exactly the same. Not sure if we need to test transactional support as a product test, since different transaction related test cases are already covered as part of an integration test.

findepi replied:
Testing that hive can correctly read the stats is a good idea.
Testing transactions -- the existing tests are detached from actual usage and they do not convince me that in `BEGIN; INSERT; SELECT` the `SELECT` will correctly see stats during planning. I don't insist however.

arhimondr replied:
Transactional support tested in `AbstractTestHiveClient`. It checks whether statistics update is visible within transaction. Whether it is not visible from outside transaction. Whether update is being committed/rollbacked successfully.

findepi commented on presto-hive/src/test/java/com/facebook/presto/hive/AbstractTestHiveClient.java:
```diff
@@ -2624,10 +2720,20 @@ private void doInsertIntoNewPartition(HiveStorageFormat storageFormat, SchemaTab
             // verify we did not modify the table directory
-            assertEquals(listAllDataFiles(transaction, tableName.getSchemaName(), tableName.getTableName()), existingFiles);
+            // TODO: After finish insert data is becoming visible in current transaction.
+            // TODO: Is this correct behaviour? Should it be consistent with inserting into existing table/partition?
```
What about this?

arhimondr replied:
It was kinda a weird behavioral. Metadata updates when inserting into an unpartitioned table are visible from within transaction, but table data is not. When for unpartitioned tables both are visible. But since https://github.com/prestodb/presto/blob/master/presto-hive/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java#L1830 it was intentional, and it is not a part of this PR to fix that.

findepi commented on presto-hive/src/test/java/com/facebook/presto/hive/AbstractTestHiveClient.java:
```diff
@@ -2624,10 +2720,20 @@ private void doInsertIntoNewPartition(HiveStorageFormat storageFormat, SchemaTab
+            // TODO: Despite partition could be get by id in current transaction, it is not returned in a partition list Oo ??
+            // Domain domain = Domain.singleValue(createUnboundedVarcharType(), utf8Slice(CREATE_TABLE_PARTITIONED_DATA_2ND_PARTITION_VALUE));
+            // TupleDomain<ColumnHandle> tupleDomain = TupleDomain.withColumnDomains(ImmutableMap.of(dsColumn, domain));
+            // TableStatistics tableStatistics = metadata.getTableStatistics(session, tableHandle, new Constraint<>(tupleDomain, bindings -> true));
```
What about this code?

arhimondr replied:
The idea is to uncomment this once we support `INSERT INTO` existing bucketed `PARTITION`/`TABLE`.

arhimondr commented on presto-hive/src/main/java/com/facebook/presto/hive/metastore/file/FileHiveMetastore.java:
```diff
@@ -399,6 +400,18 @@ public synchronized void renameTable(String databaseName, String tableName, Stri
+
+        getRequiredTable(databaseName, tableName);
+
+        alterTable(databaseName, tableName, table -> table.withParameters(parameters));
```
`org.apache.hadoop.hive.metastore.api.Table` has more fields than `com.facebook.presto.hive.metastore.Table`. Since `alterTable` has `DROP/CREATE` semantics (in other words "overwrite everyting"), when calling `alterTable` on `com.facebook.presto.hive.metastore.Table` we can overwrite some information that Presto doesn't read/write, but some other tools do.

arhimondr commented on presto-hive/src/main/java/com/facebook/presto/hive/metastore/glue/GlueHiveMetastore.java:
```diff
@@ -417,6 +417,25 @@ public void renameTable(String databaseName, String tableName, String newDatabas
+            tableInput.setParameters(parameters);
+            glueClient.updateTable(new UpdateTableRequest()
+                    .withDatabaseName(databaseName)
+                    .withTableInput(tableInput));
```
Yes. Actually we don't have any unit tests for raw `ExtendedHiveMetastore`. We rely on `AbstractTestHiveClient` to test communication with different types of metastores. Before introduction of `SemiTransactionalHiveMetastore` - `AbstractTestHiveClient` was a parent class for a kind of test you would like to see. I'm not sure adding yet another `AbstractTestHiveClient` that operates on bare `ExtendedHiveMetastore` is something worth doing, since it requires additional infra in case we want to test real HiveMetastore with docker.

arhimondr commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveBasicStatistics.java:
```diff
@@ -81,6 +83,38 @@ public String toString()
+        }
+    }
+
+    public Map<String, String> toPartitionParameters()
```
Also added tests for add/subtract

arhimondr commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java:
```diff
@@ -888,14 +895,32 @@ public HiveOutputTableHandle beginCreateTable(ConnectorSession session, Connecto
 
+        if (table.getPartitionColumns().isEmpty()) {
+            List<PartitionUpdate> unpartitionedTableUpdate = partitionUpdates.stream()
+                    .filter(u -> u.getName().isEmpty())
```
This is a concatenation of partition values. `col1=val1/col2=val2/col3=val3`

https://github.com/prestodb/presto/blob/master/presto-hive/src/main/java/com/facebook/presto/hive/HivePartitionManager.java#L288

arhimondr commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java:
```diff
@@ -888,14 +895,32 @@ public HiveOutputTableHandle beginCreateTable(ConnectorSession session, Connecto
+                parameters.putAll(update.getStatistics().toPartitionParameters());
+                table = Table.builder(table)
+                        .setParameters(parameters)
+                        .build();
```
> Would it be possible to create table object once? Maybe move this logic before Table 

I was badly trying to do so. In order to set statistic properties we need full list of `PartitionUpdate`. To get a full list of `PartitionUpdate` for a bucketed table we need to create empty files for missing buckets. To create missing files a `Table` object is required. There are a lot of methods to refactor to have another version of them that accept pretty much all the fields from `Table` but not the `Table` obejct itself. Although i agree this part of code is not nice, having another version of methods that accept 4-8 different parameters will be even more cumbersome. So i decided to leave it as is for now.

arhimondr replied:
> Also, consider extracting partitionUpdates -> update.getStatistics().toPartitionParameters() conversion to a separate method, for readability.

I exctracted that boilerplate code to `MetastoreUtil#incrementStatisticsCounters`

arhimondr commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java:
```diff
@@ -888,14 +895,32 @@ public HiveOutputTableHandle beginCreateTable(ConnectorSession session, Connecto
             }
-            partitionUpdates.forEach(partitionUpdate ->
-                    metastore.addPartition(session, handle.getSchemaName(), handle.getTableName(), buildPartitionObject(session, table, partitionUpdate), partitionUpdate.getWritePath()));
+            for (PartitionUpdate update : partitionUpdates) {
```
`table` is no longer `Effectively Immutable`.

arhimondr commented:
@findepi @kokosing Thanks for the first pass. I addressed the comments. Also i added 3 new commits

- `Store table level statistics for partitioned tables`
- `Store partitions count in table properties`
- `Update table level statistics on partition delete`

Please continue review

@haozhun Since you are the author of most of the code in `SemiTransactionalHiveMetastore` modifications to which is a most tricky part of this PR, your review will be highly appreciated.

kokosing commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveBasicStatistics.java:
```diff
@@ -0,0 +1,144 @@
+    public String toString()
+    {
+        return toStringHelper(this)
+                .add("fileCount", fileCount)
```
it would be nice to skip fields that are absent (`OptionalLong.empty()`)

arhimondr replied:
`toString` is needed here rather for a debugging purposes. Missing some fields might be rather misleading (it may seem that those fields were just forgotten to be added to the toString output). When filed is nullable, we do print `null` in `toString`. I thing in general it should be similar for `Optional`. By `in general` i mean the cases when we don't need a specific `toString` (like in `Expression`), but we need a toString for a general debugging needs.

findepi replied:
fyi you can always `com.google.common.base.MoreObjects.ToStringHelper#omitNullValues` (just fyi, no change requested)

kokosing commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveBasicStatistics.java:
```diff
@@ -0,0 +1,144 @@
+        return new HiveBasicStatistics(numFiles, numRows, inMemoryDataSizeInBytes, onDiskDataSizeInBytes);
+    }
+
+    private static OptionalLong convertStringParameter(@Nullable String parameterValue)
```
`s/convertStringParameter/parse/`?

kokosing commented on presto-hive/src/main/java/com/facebook/presto/hive/util/StatisticsUtils.java:
```diff
@@ -0,0 +1,78 @@
+        return unmodifiableMap(updatedParameters);
+    }
+
+    private static void checkAllStatisticsArePresent(HiveBasicStatistics statistics)
```
maybe you could move this method to `HiveBasicStatistics`

arhimondr replied:
Why? Do we usually do this? We rather have `check(some_obejct_to_check)` than `some_obejct_to_check.check()`. Let me know if i didn't get your point right.

arhimondr replied:
Oh, got you know =) I'm slow. Yeah, that makes sense. Will do.

arhimondr replied:
I just checked the code. I don't want to do `checkArgument(statistics.isAllStatisticsPresent)` because i want to have granular error message (like what statistic is missing).

kokosing commented on presto-product-tests/src/main/java/com/facebook/presto/tests/hive/TestHiveBasicTableStatistics.java:
```diff
@@ -0,0 +1,473 @@
+
+    private static BasicStatistics getBasicStatisticsForTable(QueryExecutor executor, String table)
+    {
+        QueryResult result = executor.executeQuery(format("DESCRIBE FORMATTED %s", table));
```
`s/executor/onHive()` this query is is not going to run with any there query exeutor

arhimondr replied:
Since there's a lot of interaction between Hive and Presto i though it will be easier to read it when you can see where the query is executed at every single step. I don't have a strong opinion though.

kokosing commented on presto-product-tests/src/main/java/com/facebook/presto/tests/hive/TestHiveBasicTableStatistics.java:
```diff
@@ -0,0 +1,473 @@
+
+    private static class BasicStatistics
+    {
+        private OptionalLong numFiles;
```
final

kokosing commented on presto-product-tests/src/main/java/com/facebook/presto/tests/hive/TestHiveBasicTableStatistics.java:
```diff
@@ -0,0 +1,473 @@
+            assertThat(statistics.getNumPartitions().getAsLong()).isEqualTo(0);
+        }
+        finally {
+            onPresto().executeQuery(format("DROP TABLE IF EXISTS %s", tableName));
```
is `onPresto().` needed? Why not use simple `io.prestodb.tempto.query.QueryExecutor.query`?

arhimondr replied:
Same explanation. I think `query` is good when you interact only with Presto.

kokosing commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java:
```diff
@@ -167,6 +167,7 @@
 import static com.google.common.collect.ImmutableMap.toImmutableMap;
```
cmt msg: s/Opensource/Open source/

kokosing commented on presto-hive/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java:
```diff
@@ -389,13 +388,47 @@ public synchronized void finishInsertIntoExistingTable(
+
+    private static <T> boolean isEmpty(Optional<List<T>> list)
+    {
+        return !list.isPresent() || list.get().isEmpty();
```
`list.map(List::isEmpty).or(true)`?

arhimondr replied:
I don't think it is more readable. But i don't have any strong opinion.

kokosing commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveBasicStatistics.java:
```diff
@@ -138,6 +138,23 @@ private static OptionalLong add(OptionalLong first, OptionalLong second)
         return OptionalLong.empty();
```
do you have a test where you drop partitions from external table which did not have any statistics?

arhimondr replied:
Yes, a unit and an integration.

kokosing commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveBasicStatistics.java:
```diff
@@ -138,6 +138,23 @@ private static OptionalLong add(OptionalLong first, OptionalLong second)
         return OptionalLong.empty();
     }
```
can you add product test for that?

findepi commented on presto-hive/src/main/java/com/facebook/presto/hive/util/StatisticsUtils.java:
```diff
@@ -53,18 +57,35 @@ public static Partition incrementStatisticsCounters(Partition partition, HiveBas
         updatedParameters.putAll(updatedStatistics.toPartitionParameters());
+        if (partitions.isPresent()) {
+            OptionalInt currentPartitionCount = getPartitionCount(parameters);
+            currentPartitionCount.ifPresent(count -> updatedParameters.put(NUM_PARTITIONS, Integer.toString(count + partitions.getAsInt())));
```
https://cwiki.apache.org/confluence/display/Hive/StatsDev refers to "Fictional metastore property: `numPartitions`". Is this supposed to be a read-only table property?

arhimondr replied:
Yeah. But until we are on Hive 2.3+ we need to maintain a non-fictional property.

findepi replied:
Ok. And what happens in Hive 2.3+? Does it complain when you set this? (I know product tests use Hive >= 2, not sure which exactly)

findepi commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java:
```diff
@@ -667,6 +668,8 @@ private void createHiveTable(ConnectorSession session, ConnectorTableMetadata ta
 
         tableProperties.putAll(createZeroStatistics().toPartitionParameters());
 
+        tableProperties.put(NUM_PARTITIONS, "0");
```
Should we set this property for non-partitioned tables?

arhimondr replied:
No, good point

findepi commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java:
```diff
@@ -154,6 +154,7 @@
 import static com.facebook.presto.hive.metastore.StorageFormat.VIEW_STORAGE_FORMAT;
 import static com.facebook.presto.hive.metastore.StorageFormat.fromHiveStorageFormat;
 import static com.facebook.presto.hive.util.ConfigurationUtils.toJobConf;
+import static com.facebook.presto.hive.util.StatisticsUtils.NUM_PARTITIONS;
```
> Store partitions count in table properties

Same general comments as in `Store table level statistics for partitioned tables`:
- can affect concurrency
- can go out of sync if there are other writers to metastore

And also:
- https://cwiki.apache.org/confluence/display/Hive/StatsDev looks as if this was not supposed to be set

arhimondr replied:
> can affect concurrency

It does not affect concurrency right now. Actually there is no locking in SemiTransactionalHiveMetastore that affects multiple transactions.

But the problem is that, yeah

> can go out of sync if there are other writers to metastore

I'm going to change `updateTableParameter` methods to be `compareAndSet` to do not loose any updates.

findepi commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java:
```diff
@@ -1307,16 +1307,26 @@ public OptionalLong metadataDelete(ConnectorSession session, ConnectorTableHandl
             }
+            metastore.finishDeleteFromExistingTable(session, handle.getSchemaName(), handle.getTableName(), deletedDataStatistics, deletedPartitionsCount);
         }
-        // it is too expensive to determine the exact number of deleted rows
```
is it no longer expensive? :)

Does it mean that `DELETE FROM hive.table WHERE partkey = 1` will now report number of deleted rows? (But it will be stats-based, potentially inexact?)

arhimondr replied:
Yes

findepi commented on presto-hive/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java:
```diff
@@ -1776,7 +1856,8 @@ private void checkHoldsLock()
         ALTER,
-        INSERT_EXISTING
+        INSERT_EXISTING,
+        DELETE_EXISTING
```
nit: add `,` to avoid diffs like this one

findepi commented on presto-hive/src/main/java/com/facebook/presto/hive/util/StatisticsUtils.java:
```diff
@@ -0,0 +1,78 @@
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Collections.unmodifiableMap;
+
+public class StatisticsUtils
```
`final` (otherwise a warning in intellij)

findepi commented on presto-hive/src/main/java/com/facebook/presto/hive/util/StatisticsUtils.java:
```diff
@@ -16,15 +16,19 @@
 
 public class StatisticsUtils
 {
+    public static final String NUM_PARTITIONS = "numPartitions";
```
nit: I would keep `private StatisticsUtils() {}` first, since it's not a real constructor, just a utility class marker

findepi commented on presto-product-tests/src/main/java/com/facebook/presto/tests/TestGroups.java:
```diff
@@ -61,6 +61,7 @@
     public static final String TLS = "tls";
     public static final String CANCEL_QUERY = "cancel_query";
     public static final String BIG_QUERY = "big_query";
+    public static final String HIVE_TABLE_STATISTICS = "hive_table_statistics";
```
i don't think we need this test group (but i don't have strong opinion)

arhimondr replied:
Unfortunately it is problematic to run the tests by a class name (at least as i know). I added it to be able to run only this subset of tests. @kokosing can we run a subset of tests by specifying a class name?

findepi commented on presto-product-tests/src/main/java/com/facebook/presto/tests/hive/TestHiveBasicTableStatistics.java:
```diff
@@ -0,0 +1,473 @@
+        try {
+            BasicStatistics statistics = getBasicStatisticsForTable(onHive(), tableName);
+            assertThatStatisticsArePresent(statistics);
+            assertThatStatisticsAreNonZero(statistics);
```
`assertThatStatisticsAreNonZero` should include `assertThatStatisticsArePresent` (and then you don't need to call it explicitly)

findepi commented on presto-product-tests/src/main/java/com/facebook/presto/tests/hive/TestHiveBasicTableStatistics.java:
```diff
@@ -0,0 +1,473 @@
+        return createFromDescribeResult(result);
+    }
+
+    private static BasicStatistics createFromDescribeResult(QueryResult result)
```
`basicStatisticsFromDescribeResult`

findepi commented on presto-product-tests/src/main/java/com/facebook/presto/tests/hive/TestHiveBasicTableStatistics.java:
```diff
@@ -0,0 +1,473 @@
+            // Insert into bucketed unpartitioned table is unsupported
+            // Once supported - uncomment assertions below
+            assertThatThrownBy(() -> insertNationData(onPresto(), tableName))
+                    .hasMessageContaining("Cannot insert into bucketed unpartitioned Hive table");
```
❤️

findepi commented on presto-product-tests/src/main/java/com/facebook/presto/tests/hive/TestHiveBasicTableStatistics.java:
```diff
@@ -0,0 +1,473 @@
+            // BasicStatistics statisticsAfterInsert = getBasicStatisticsForTable(onHive(), tableName);
+            // assertThatStatisticsValuesHaveIncreased(statisticsAfterCreate, statisticsAfterInsert);
+            // assertThat(statisticsAfterInsert.getNumRows().getAsLong()).isEqualTo(50);
+            // assertThat(statisticsAfterCreate.getNumFiles().getAsLong()).isEqualTo(100);
```
Technically, you could uncomment these lines (with expected values := un-updated values)

findepi commented on presto-product-tests/src/main/java/com/facebook/presto/tests/hive/TestHiveBasicTableStatistics.java:
```diff
@@ -0,0 +1,473 @@
+        String tableName = "test_basic_statistics_unpartitioned_ctas_presto";
+
+        onPresto().executeQuery(format("DROP TABLE IF EXISTS %s", tableName));
+        onPresto().executeQuery(format("CREATE TABLE %s AS SELECT * FROM nation", tableName));
```
Please add a test for `WITH NO DATA`

arhimondr commented on presto-hive/src/test/java/com/facebook/presto/hive/HiveBenchmarkQueryRunner.java:
```diff
@@ -65,7 +66,7 @@ public static LocalQueryRunner createLocalQueryRunner(File tempDir)
 
         // add hive
         File hiveDir = new File(tempDir, "hive_data");
-        TestingHiveMetastore metastore = new TestingHiveMetastore(hiveDir);
```
?

findepi replied:
Before, `hiveDir + "/metastore"` was used as a metastore dir, not it's directly `hiveDir`. Perhaps this doesn't matter though

arhimondr replied:
Got you. Yeah, i don't think it does matter.

arhimondr commented on presto-hive/src/test/java/com/facebook/presto/hive/metastore/thrift/TestingHiveMetastore.java:
```diff
@@ -1,783 +0,0 @@
-import static org.apache.hadoop.hive.metastore.TableType.VIRTUAL_VIEW;
-
-@ThreadSafe
-public class TestingHiveMetastore
```
TestingHiveMetastore had nothing to do with thrift, have no idea why did it end up in that package. So i just did `s/TestingMetastore/TestingHiveMetastore/`

arhimondr commented on presto-hive/src/main/java/com/facebook/presto/hive/RecordFileWriter.java:
```diff
@@ -122,6 +125,15 @@ public long getWrittenBytes()
 
+        if (committed) {
+            try {
+                return path.getFileSystem(conf).getFileStatus(path).getLen();
```
`ExtendedRecordWriter` reports `writtenBytes` on the go. I would leave it as is.

findepi replied:
Yes, i noticed. Why not improve default record writer to report written bytes on the go as well? Still, according to the commit message that introduced the `ExtendedRecordWriter`, the main difference lied in `path.getFileSystem(conf).getFileStatus(path).getLen();` -- which is no longer a difference.

arhimondr replied:
> Why not improve default record writer to report written bytes on the go as well? 

DefaultWriter uses writers from Hive. Not all of them report the writen bytes statistic.

arhimondr replied:
Oh, i thought `ExtendedRecordWriter` is a Hive class. Yeah, let me check that.

arhimondr replied:
So, basically we use this for `RCFile.Writer` because specifically `RCFile.Writer` provides `#getLength()` method that can be used to get bytes written. We do `Object writer = Class.forName(outputFormatName).getConstructor().newInstance();` when getting all the other writers. So we would have to do a big switch instead and apply `ExtendedFileWriter` whenever it is possible. Although we got not that many supported formats, it still feels to be tedious.

arhimondr commented on presto-hive/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java:
```diff
@@ -1060,12 +1084,22 @@ private void prepareAddPartition(HdfsContext context, PartitionAndMore partition
+            Partition oldPartition = delegate.getPartition(partition.getDatabaseName(), partition.getTableName(), partition.getValues())
+                    .orElseThrow(() -> new PrestoException(
+                            TRANSACTION_CONFLICT,
+                            format("The partition that this transaction modified was deleted in another transaction. %s %s", partition.getTableName(), partition.getValues())));
```
This is good to avoid having placeholders within a message. It is easier to grep error messages through the code than. Added labels though.

arhimondr replied:
You don't need additional punctuation as well. `Collection#toString` does the thing.

arhimondr commented on presto-hive/src/test/java/com/facebook/presto/hive/AbstractTestHiveClient.java:
```diff
@@ -2346,6 +2367,13 @@ protected void doCreateTable(SchemaTableName tableName, HiveStorageFormat storag
+            assertEquals(statistics.getRowCount().getAsLong(), CREATE_TABLE_DATA.getRowCount());
+            assertEquals(statistics.getFileCount().getAsLong(), 1L);
+            assertGreaterThan(statistics.getInMemoryDataSizeInBytes().getAsLong(), 0L);
+            assertGreaterThan(statistics.getOnDiskDataSizeInBytes().getAsLong(), 0L);
```
I'm afraid we can run out of value for some exotic storage format / table data, and than someone will get very frustrating test failure. Let's just rely on the fact that 3+ engineers have looked at this code, and there is no typo.

findepi replied:
:) If some new or exotic data format can compress the data below <some-non-stupid-fix-point>, they they will get a 🏅 in computer science.

arhimondr replied:
I mean we can chose this value to be small enough. By why 120? Why not 119? If we chose something closer to 0 than we got the same problem (something close to 0 might a files count). If we chose something bigger (like `100+`) there is a chance that some of the compression algorithms one day can compress the data to become `99`. And than i don't want to be an engineer who decided to `Slightly improve serialization of blahblah`

arhimondr commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java:
```diff
@@ -1097,35 +1094,56 @@ public HiveInsertTableHandle beginInsert(ConnectorSession session, ConnectorTabl
+            verify(partitionUpdates.size() <= 1, "partitionUpdates are expected to be merged at this point");
+            Optional<PartitionUpdate> partitionUpdate = partitionUpdates.stream().collect(toOptional());
+            if (partitionUpdate.isPresent()) {
+                PartitionUpdate update = partitionUpdate.get();
```
Yeah, indeed =))

arhimondr commented on presto-hive/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java:
```diff
@@ -389,13 +388,47 @@ public synchronized void finishInsertIntoExistingTable(
+                    return;
+                }
                 throw new UnsupportedOperationException("Inserting into an unpartitioned table that were added, altered, or inserted into in the same transaction is not supported");
+            }
```
Yeah, but i prefer to put curlies when there is more than a one line in a switch case for readability.

arhimondr commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java:
```diff
@@ -609,7 +610,7 @@ private void createHiveTable(ConnectorSession session, ConnectorTableMetadata ta
         List<HiveColumnHandle> columnHandles = getColumnHandles(tableMetadata, ImmutableSet.copyOf(partitionedBy), typeTranslator);
         HiveStorageFormat hiveStorageFormat = getHiveStorageFormat(tableMetadata.getProperties());
-        Map<String, String> tableProperties = getEmptyTableProperties(tableMetadata, !partitionedBy.isEmpty());
+        Map<String, String> tableProperties = getEmptyTableProperties(tableMetadata);
```
> What happens if two transactions try to add a new partition to an existing table? This should fail since there is no place where we could mediate the final result, right?

There is. Read `SemiTransactionalHiveMetastore#finishInsertIntoExistingTable`

arhimondr replied:
> "Open source hive does not maintain table level statistics for partitioned table", so the table-level stats are going to be off-sync when e.g. Hive is used to create some partitions. Or ?

It is highly unlikely that you would create a table from `Presto` (that's the only way those statistics could be maintained by Presto on INSERT than) and than insert into that table from Hive. Technically we can add a feature flag, but i don't think it is worth additional code for testing.

arhimondr commented on presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java:
```diff
@@ -888,14 +894,23 @@ public HiveOutputTableHandle beginCreateTable(ConnectorSession session, Connecto
+        if (table.getPartitionColumns().isEmpty()) {
+            HiveBasicStatistics tableStatistic = partitionUpdates.stream()
+                    .map(PartitionUpdate::getStatistics)
+                    .reduce(HiveBasicStatistics::add)
```
It can be bucketed. Than we will have 1 update per bucket.

arhimondr commented:
@findepi @kokosing 

I had a discussion with @haozhun yesterday. Turns out we don't have to store table statistics for partitioned tables. Since this was not compatible with the open source Hive anyway, I'm dropping out the next commits:

- https://github.com/prestodb/presto/pull/10456/commits/26a0547cb33c16a0842d089992df02ca6f360990 (Store table level statistics for partitioned tables)
- https://github.com/prestodb/presto/pull/10456/commits/55ab09a7c14e0a02620fc939592348754e4cf362 (Store partitions count in table properties)
- https://github.com/prestodb/presto/pull/10456/commits/644e5c9fad5455d44e1eeb1438e637e10dd5a69b (Update table level statistics on partition delete)

We discussed the options of how to preserve the statistics consistent (in other words how to avoid race conditions when updating stats in the partition parameters).

It is decided to change the `ExtenendedHiveMetastore#replaceParameters(... Map newParameters)` method to the `ExtenendedHiveMetastore#updateParameters(... Function<Map, Map> updater)`.

This makes it possible to implement an "atomic" update for a metastore.

As the initial implementation i made the `update*` methods to be synchronized. That will ensure atomicity for the case when a single Presto cluster has an exclusive control over a table. However it obviously does not solve the problem for the rest of the cases. 

The better implementation would require modifications in the HiveMetastore code. We need the HiveMetastore to expose `compareAndSet` kind methods for both a `Table` and a `Partition` (or at least for a `Table` and a `Partition` parameters). Since this is something we cannot do in the very near future, we decided to go with the current implementation for now.

haozhun commented on presto-hive/src/main/java/com/facebook/presto/hive/PartitionUpdate.java:
```diff
@@ -34,28 +35,52 @@
+        this.rowCount = rowCount;
+        checkArgument(inMemoryDataSizeInBytes >= 0, "inMemoryDataSizeInBytes is expected to be greater than or equal to zero: %d", inMemoryDataSizeInBytes);
+        this.inMemoryDataSizeInBytes = inMemoryDataSizeInBytes;
+        checkArgument(onDiskDataSizeInBytes >= 0, "onDiskDataSizeInBytes is expected to be greater than or equal to zero: %d", onDiskDataSizeInBytes);
```
Ditto about message for these `checkArguments`

haozhun commented on presto-hive/src/main/java/com/facebook/presto/hive/RecordFileWriter.java:
```diff
@@ -122,6 +125,15 @@ public long getWrittenBytes()
+            }
+        }
+
         // there is no good way to get this from RecordWriter
```
The comment is now somewhat misleading. Please update.

arhimondr replied:
Don't know how to rephrase it without being oververbose. Any help is appreciated.

haozhun commented on presto-hive/src/main/java/com/facebook/presto/hive/metastore/SemiTransactionalHiveMetastore.java:
```diff
@@ -2193,6 +2307,62 @@ public void undo(ExtendedHiveMetastore metastore)
 
+    private static class UpdateStatisticsOperation
+    {
+        private static final int MAX_ATTEMPTS = 100;
```
unused

haozhun commented on presto-hive/src/main/java/com/facebook/presto/hive/metastore/Table.java:
```diff
@@ -106,6 +106,11 @@ public String getTableType()
         return partitionColumns;
     }
 
+    public boolean isPartitioned()
```
I recommend inlining this method. There are already quite a few places where we do `.getPartitionColumns().isEmpty` directly

haozhun commented on presto-product-tests/src/main/java/com/facebook/presto/tests/hive/TestHiveBasicTableStatistics.java:
```diff
@@ -0,0 +1,404 @@
+        }
+    }
+
+    private static void insertNationData(QueryExecutor executor, String tableName)
```
I think it makes sense to make `insertNationData` ,`getBasicStatisticsForTable`, and `getBasicStatisticsForPartition` not static so that they can call `onHive` and `onPresto` directly.

That will be easier to read.

But I don't feel strongly. Up to you.

haozhun commented:
I have some comments. All of them are pretty minor.

Looks good, overall.

findepi commented on presto-hive/src/test/java/com/facebook/presto/hive/metastore/thrift/InMemoryHiveMetastore.java:
```diff
@@ -572,6 +575,16 @@ private static boolean isParentDir(File directory, File baseDirectory)
         return false;
     }
 
+    private static void deleteDirectory(File dir)
```
`InMemoryHiveMetastore` sounds like "keep things in memory", not on disk. Maybe, this is actually so because metadata is in memory and data is on disk, which wouldn't be sad.

arhimondr replied:
This is the weird thing about the Hive Metastore. I touches the data "almost" never. The metastore does that only for `DROP TABLE` and `DROP PARTITION` operations. It removes the managed data. In `InMemoryHiveMetastore` we want to simulate the same semantics.

arhimondr commented on presto-hive/src/main/java/com/facebook/presto/hive/PartitionStatistics.java:
```diff
@@ -18,70 +18,37 @@
-            Map<String, HiveColumnStatistics> columnStatistics)
+            HiveBasicStatistics basicStatistics,
+            Map<String, HiveColumnStatistics> columnStatistics,
+            boolean columnStatsAccurate)
```
This property is used internally by Hive to execute `SELECT max(column)` kind of queries efficiently. If the statistics are accurate, this query won't scan any data, and will return max value for a give column using statistics.


The thing is that this field is no longer boolean. There is a JSON in it (can't say exactly from what Hive version though). That JSON is a map, that has `boolean basicStatisticsAccurate` and `Map<Column, Boolean> columnStatisticsAccurate` fields. Hive set's those values to `TRUE` after Analyze is run, and it set's those value to `FALSE` if any data modification occurred. 

In Presto we don't do any optimizations of such kind, so we can simply ignore this field (ideally we had to set it to `FALSE` after an INSERT operation, but since we update statistics now - we can just leave it untouched)

arhimondr replied:
I think i'm just gonna remove this field from our codebase.

findepi commented:
> @arhimondr merged 

🎉
